.github/workflows/preview-docs.yml

```diff
@@ -1,28 +1,105 @@
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          PR_NUMBER: ${{ github.event.workflow_run.pull_requests[0].number }}
+        run: |
+          python dev/preview_docs.py \
+            --commit-sha ${{ github.event.workflow_run.head_sha }} \
+            --pull-number $PR_NUMBER \
+            --workflow-run-id ${{ github.run_id }} \
+            --docs-workflow-run-url ${{ github.event.workflow_run.html_url }} \
+            --stage requested
+
+  fail-early:
+    if: github.repository == 'mlflow/mlflow' && github.event.workflow_run.status == 'completed'
+    runs-on: ubuntu-latest
+    permissions:
+      pull-requests: write # preview_docs.py comments on PRs
+    timeout-minutes: 10
+    steps:
+      - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
+        if: github.event.workflow_run.conclusion == 'failure'
+      - uses: ./.github/actions/setup-python
+        if: github.event.workflow_run.conclusion == 'failure'
+      - name: Install dependencies
+        if: github.event.workflow_run.conclusion == 'failure'
+        run: |
+          pip install requests
+      - name: Handle docs workflow failure
+        if: github.event.workflow_run.conclusion == 'failure'
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          PR_NUMBER: ${{ github.event.workflow_run.pull_requests[0].number }}
+        run: |
+          python dev/preview_docs.py \
+            --commit-sha ${{ github.event.workflow_run.head_sha }} \
+            --pull-number $PR_NUMBER \
+            --workflow-run-id ${{ github.run_id }} \
+            --docs-workflow-run-url ${{ github.event.workflow_run.html_url }} \
+            --stage failed
+          exit 1
+
   main:
-    if: github.repository == 'mlflow/mlflow'
+    needs: [fail-early]
+    if: needs.fail-early.outputs.status == 'success'
     runs-on: ubuntu-latest
     permissions:
-      statuses: read # preview_docs.py checks PR statuses
       pull-requests: write # preview_docs.py comments on PRs
     timeout-minutes: 10
     steps:
       - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
+      - name: Download build artifact
+        uses: actions/download-artifact@95815c38cf2ff2164869cbab79da8d1f422bc89e # v4.2.1
+        with:
+          name: docs-build-${{ github.event.workflow_run.id }}
+          path: downloaded-artifact
       - uses: ./.github/actions/setup-python
+      - uses: ./.github/actions/setup-node
       - name: Install dependencies
         run: |
           pip install requests
+          npm install -g netlify-cli
+      - name: Deploy to Netlify
+        id: netlify_deploy
+        working-directory: downloaded-artifact
+        env:
+          NETLIFY_AUTH_TOKEN: ${{ secrets.NETLIFY_AUTH_TOKEN }}
+          NETLIFY_SITE_ID: ${{ secrets.NETLIFY_SITE_ID }}
+          PR_NUMBER: ${{ github.event.workflow_run.pull_requests[0].number }}
+        run: |-
+          netlify deploy \
```

Comment:
This is another follow-up item. I wonder if it's possible to serve docs in `https:://...com/docs/latest/` to minimize the difference between prod and preview.

Reply:
Noted!